### PR TITLE
Support instanceName on HasId without a component by using localName.

### PR DESCRIPTION
**Pull request by mikeurbach**
### Contributor Checklist

- [ ] Did you add Scaladoc to every public function/method?
- [x] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [ ] Did you add appropriate documentation in `docs/src`?
- [x] Did you request a desired merge strategy?
- [x] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement

- Feature (or new API)

#### Desired Merge Strategy

- Squash: The PR will be squashed and merged (choose this if you have no preference).

#### Release Notes

We have recently reworked when HasIds are named within a module so they can be named before atModuleBodyEnd executes, and then get referenced within atModuleBodyEnd. However, the component is still not set when they are referenced, so in this case use the local name.

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels? (Select the most appropriate one based on the "Type of Improvement")
- [ ] Did you mark the proper milestone (Bug fix: `3.5.x`, `3.6.x`, or `5.x` depending on impact, API modification or big change: `6.0`)?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you do one of the following when ready to merge:
  - [ ] Squash: You/ the contributor `Enable auto-merge (squash)`, clean up the commit message, and label with `Please Merge`.
  - [ ] Merge: Ensure that contributor has cleaned up their commit history, then merge with `Create a merge commit`.

### CODE REVIEW

**Review by azidar — APPROVED** [review]

File: core/src/main/scala/chisel3/internal/Builder.scala [hunk 1]
```diff
@@ -315,6 +315,13 @@ private[chisel3] trait HasId extends chisel3.InstanceId {
         case (Some(c), _) => refName(c)
         case (None, d: Data) if d.topBindingOpt == Some(CrossModuleBinding) => _ref.get.localName
         case (None, _: MemBase[_]) => _ref.get.localName
+        case (None, _) if _ref.isDefined => {
+          // Support instance names for HasIds that don't have a _parent set yet, but do have a _ref set.
+          // This allows most HasIds to be named in atModuleBodyEnd, for example.
+          // However, this is only valid if the _ref's fullName doesn't depend on the component name.
+          // Notably, this is not valid for ports of instances, and other constructs that need the _parent's name.
```
Comment (jackkoenig) [inline core/src/main/scala/chisel3/internal/Builder.scala:322]: This comment is a little bit concerning, what happens if you use `.toRelativeTarget` on a port of a submodule?

Comment (mikeurbach) [reply core/src/main/scala/chisel3/internal/Builder.scala:322]: I think I may have made this comment sound too scary, without thinking deeply enough about it. I just pushed up a commit with a submodule, is this what you had in mind? [68a31cb](https://github.com/chipsalliance/chisel/pull/3623/commits/68a31cb4a8efd6d6027344ca9da7762d03b03d40)

I think we actually are fine here. If we're trying to name a port of a submodule, the _parent component of the port will always be set, so we wouldn't come through this code path anyway. The only time we should come here is if we're naming a port of the currently-being-elaborated module, in which case the localName is safe to use (I also added a test demonstrating that).

Does this make sense? Maybe I just need to amend the comment to indicate that in fact any time we hit this code path, it's safe to use a localName.

Comment (mikeurbach) [reply core/src/main/scala/chisel3/internal/Builder.scala]: FYI I've pushed another commit that updates the comment based on what I just said here.

Comment (jackkoenig) [reply core/src/main/scala/chisel3/internal/Builder.scala]: Yep makes sense, thank you for the test and clarification!

**Review by jackkoenig — APPROVED** [review]

**Comment by mikeurbach** [comment]
Thanks for the reviews!

**Pull request merged**